examples/dishwasher-app/silabs/src/AppTask.cpp

```diff
@@ -0,0 +1,237 @@
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <assert.h>
+
+#include <app-common/zap-generated/attribute-type.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app/server/OnboardingCodesUtil.h>
+#include <app/server/Server.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/silabs/platformAbstraction/SilabsPlatform.h>
+#include <setup_payload/QRCodeSetupPayloadGenerator.h>
+#include <setup_payload/SetupPayload.h>
+
+#include "AppConfig.h"
+#include "AppEvent.h"
+#include "AppTask.h"
+#include "DeviceEnergyManager.h"
+#include "DishwasherManager.h"
+#include "ElectricalSensorManager.h"
+#include "operational-state-delegate-impl.h"
+
+#include <app/clusters/device-energy-management-server/DeviceEnergyManagementTestEventTriggerHandler.h>
+#include <app/clusters/electrical-energy-measurement-server/EnergyReportingTestEventTriggerHandler.h>
+
+#define APP_FUNCTION_BUTTON 0
+#define APP_CONTROL_BUTTON 1
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::OperationalState;
+using namespace chip::app::Clusters::DeviceEnergyManagement;
+using namespace chip::DeviceLayer;
+using namespace chip::DeviceLayer::Silabs;
+using namespace chip::TLV;
+
+AppTask AppTask::sAppTask;
+
+#if SL_MATTER_TEST_EVENT_TRIGGER_ENABLED
+static EnergyReportingTestEventTriggerHandler sEnergyReportingTestEventTriggerHandler;
+static DeviceEnergyManagementTestEventTriggerHandler sDeviceEnergyManagementTestEventTriggerHandler;
+#endif
+
+CHIP_ERROR AppTask::Init()
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::DeviceLayer::Silabs::GetPlatform().SetButtonsCb(AppTask::ButtonEventHandler);
+
+#ifdef DISPLAY_ENABLED
+    GetLCD().Init((uint8_t *) "Dishwasher-App");
+#endif
+
+    err = BaseApplication::Init();
+    if (err != CHIP_NO_ERROR)
+    {
+        SILABS_LOG("BaseApplication::Init() failed");
+        appError(err);
+    }
+
+    PlatformMgr().LockChipStack();
+    err = DeviceEnergyManager::Instance().Init();
+    PlatformMgr().UnlockChipStack();
+    if (err != CHIP_NO_ERROR)
+    {
+        SILABS_LOG("DeviceEnergyMgr.Init() failed");
+        appError(err);
+    }
+
+    PlatformMgr().LockChipStack();
+    err = ElectricalSensorManager::Instance().Init();
+    PlatformMgr().UnlockChipStack();
+    if (err != CHIP_NO_ERROR)
+    {
+        SILABS_LOG("ElectricalSensorMgr.Init() failed");
+        appError(err);
+    }
+
+#if SL_MATTER_TEST_EVENT_TRIGGER_ENABLED
+    // Register DEM & Electrical Sensor Test Event Trigger
```

Comment:
Rather than being application specific, could these be in a common Silabs platform could with #ifdefs based on Application example type - e.g. if you had a Water Heater or EVSE build then these would be enabled differently.

It depends on how silicon manufacturers want to support examples I suppose, but from a maintenance perspective from the viewpoint of someone who wants the same code to support all platforms and all applications, this approach seems to be harder to maintain.

Given you have examples/dishwasher-app/ and presumably examples/evse-app and examples water-heater-app etc then this code is going to be very similar in each one, and cut and paste each time?

I can see from the perspective of someone who only wants to make a dishwasher (and not an EVSE) then these #ifdefs are a distraction, but keeping them all in line with each other seems like more work?

Reply:
There is quite a lot of rework that needs to be done to refactor this app in a way that removing the energy management is seamless (Electrical Sensor and Device Energy Manager need to be gated by some ifdef, Dishwasher manager also now inherits DEMManufacturerDelegate so it needs some gating, etc.) so I would worry to much about supporting a non EVSE dishwasher app for now. We might update the app in the future if there is a need for it.